pkg/scheduler/queue/queue_test.go

```diff
@@ -329,6 +329,124 @@ func queueConsume(
+		// or else StopAndAwaitTerminated will never complete.
+		queue.SubmitUnregisterQuerierWorkerConn(querier2Conn)
+		require.NoError(t, services.StopAndAwaitTerminated(ctx, queue))
+	})
+
+	// Enqueue a request from a user which would NOT be assigned to querier-1.
+	// NOTE: "user-1" shuffle shard always chooses the first querier ("querier-1" in this case)
+	// when there are only one or two queriers in the sorted list of connected queriers
+	reqNotSharderToQuerier1 := &SchedulerRequest{
+		Ctx:                       context.Background(),
+		Request:                   &httpgrpc.HTTPRequest{Method: "GET", Url: "/hello"},
+		AdditionalQueueDimensions: randAdditionalQueueDimension(""),
+	}
+	// This will sit in the queue as querier-2 is not await a request yet,
+	// but we need a request in the queue tree to reach the code path
+	// in QuerierWorkerQueuePriorityAlgo.setup where the panic condition is created
+	// by setting the currentNodeOrderIndex to -1.
+	require.NoError(t, queue.SubmitRequestToEnqueue("user-2", reqNotSharderToQuerier1, 1, nil))
+
+	// Querier-1 submits a request to dequeue;
+	// it will not receive anything as the only request in the queue is not sharded to querier-1.
+	querier1SubmitDequeueReqWG := sync.WaitGroup{}
+	querier1SubmitDequeueReqWG.Add(1)
+	querier1DequeueReq := NewQuerierWorkerDequeueRequest(querier1Conn, FirstTenant())
+	go func() {
+		defer querier1SubmitDequeueReqWG.Done()
+		_, _, err = queue.AwaitRequestForQuerier(querier1DequeueReq)
+		// This receives one of two errors, it's a race between a context cancel and ErrQuerierWorkerDisconnected
+		require.Error(t, err)
+	}()
+
+	querier1CrashWG := sync.WaitGroup{}
+	querier1CrashWG.Add(1)
+	go func() {
+		defer querier1CrashWG.Done()
+		for queue.waitingDequeueRequestsToDispatch.Len() < 1 {
+			time.Sleep(time.Millisecond * 1)
+		}
+		// Simulate a crash of querier-1 (no graceful shutdown notification).
+		querier1CtxCancel(errors.New("crash"))
+		queue.SubmitUnregisterQuerierWorkerConn(querier1Conn) // normally done in the defers of the grpc loop
+	}()
+
+	// Enqueue another request from a user which would NOT be assigned to querier-1.
+	reqNotSharderToQuerier1 = &SchedulerRequest{
+		Ctx:                       context.Background(),
+		Request:                   &httpgrpc.HTTPRequest{Method: "GET", Url: "/hello"},
+		AdditionalQueueDimensions: randAdditionalQueueDimension(""),
+	}
+	// The enqueuing of any new request triggers an attempt to dispatch requests
+	// to the items in waitingDequeueRequestsToDispatch list and their associated worker conns.
+	//
+	// The dequeue request at the front of that list will be associated with the disconnected worker conn,
+	// which now has a workerID set to unregisteredWorkerID = -1.
+	// Without any check in place to bail out before dequeuing for a workerID with -1,
+	// the QuerierWorkerQueuePriorityAlgo.dequeueSelectNode would panic with index out of range [-1].
```

Comment:
```suggestion
	// Attempting to dequeue a request from the tree for an unregisteredWorkerID should fail, but not panic.
```

We don't have to explain the details of a panic that _shouldn't_ happen, we can link the bug report to this PR instead for posterity/documentation if someone wants to know.